src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -851,6 +851,22 @@ public INamedTypeSymbol GetTypeByMetadataName(string fullyQualifiedMetadataName)
         /// </summary>
         public abstract ImmutableArray<Diagnostic> GetDiagnostics(CancellationToken cancellationToken = default(CancellationToken));
 
+        internal void EnsureCompilationCompleted()
+        {
+            Debug.Assert(this.EventQueue != null);
+
+            lock (this.EventQueue)
```

Comment:
The existing ones (for non-zero tree count) do ensure this under a lock, but the new callsite in GetDiagnostics doesn't take a lock.
Let me separate this into 2 methods:

1. CompleteCompilationEventQueue_NoLock which is expected to be called with a lock (and can be called for non-zero tree count case) and
 2. EnsureCompilationEventQueueComplete: Which takes the event queue lock and calls the above method. This method can be called for the newly added zero syntax tree case.